### PR TITLE
Resim bundles

### PR DESCRIPTION
**Pull request by dvush**
## 📝 Summary

Each bundle is re-simulated for each block in the range even when simulation is successful. Bundle is sent to builder for each block in the range when its valid. This enables "soft" bundle cancellations. 

Cancellations are implemented as redis kv store with expiring keys that are written after the bundle is cancelled in the db.

* db is changes to record gas used for simulation bundle across all simulation attempts
* bundles are sent for one block only
* bundles are inserted for builder also for one block only
* db tests are added 

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test`
* [x] `go mod tidy`
* [x] I have seen and agree to `CONTRIBUTING.md`

### CODE REVIEW

**Comment by Ruteri** [comment]
What impact do you anticipate on the performance?
Does this change introduce additional latency that could negatively impact builders?

**Comment by dvush** [comment]
>What impact do you anticipate on the performance?

so there are two parts:
* for the first simulation of the bundle on the first valid block the only overhead would be time to get cancellation status from redis. Normally its around 1 ms or less for managed redis in aws.
First sim latency is important because a lot of bundles target the next block.
* for the next simulations of the same bundles impact is even less noticeable because for subsequent blocks you have the whole 12s slot time to process bundle and send it to the builder

>We would also increase dependency on the simulation cluster, since we'd no longer forward bundles for any blocks where the sim cluster was lagging

If our whole sim cluster can't process simulations we would not forward bundle before and the same situation happens here for the first simulation. 

What this resimulation adds is a possible problem when cluster is not lagging for some blocks but lagging for another so we can miss those blocks. But we would already miss bundles that were targeting those lagging blocks as their first simulation so this problem was already there, now bundles that were simulated before would also be affected by that. 

I think that this is tolerable because alternative can be to simulate only once and then forward for all subsequent blocks. This increases amount of garbage that we send to other builders because bundle will likely fail, also this system with resimulation for each block naturally handles cases when bundle(or any part of it) is included onchain. 

But what we also add as a dependency is query to redis that checks cancellations on the critical path. When it takes 1ms its alright but if redis is to fail in a mode when it hangs it can add up to 1s to the latency. On one hand 1s is too much for this case and should make this param lower and configurable. 

But on the other hand the whole queue is built around redis and if it hangs the whole queue hangs, its not perfect but I like the convenience of redis for the queue. I expect minimal possible downtime to be in minutes(5-10) because we would see alert that mevshare is down and we can quickly swap mevshare to another instance. Alternatively we can make it more automatic.

File: mevshare/database.go [hunk 1]
```diff
@@ -34,21 +34,41 @@ type DBSbundle struct {
 	SimProfit          sql.NullString `db:"sim_profit"`
 	SimRefundableValue sql.NullString `db:"sim_refundable_value"`
 	SimGasUsed         sql.NullInt64  `db:"sim_gas_used"`
-	Body               []byte         `db:"body"`
-	BodySize           int            `db:"body_size"`
-	OriginID           sql.NullString `db:"origin_id"`
+	// sum of all simulations gas used
+	SimAllSimsGasUsed sql.NullInt64 `db:"sim_all_sims_gas_used"`
```
Comment (Wazzymandias) [inline mevshare/database.go:38]: suggestion (non-blocking): We should probably update the notion doc with the new schema fields so data team is aware

cc: @elenahoo @ababino @sui414

**Review by Wazzymandias — APPROVED** [review]
Overall looks good to me. The one thing I'm concerned about is if there is a small reorg that causes the service to consider a bundle to be cancelled when it's still within its time window (block window) post-reorg. Maybe I am mistaken, and it's not that important

**Pull request merged**